File: services/agent/src/credits.ts
export const CREDIT_COSTS = {
  kex_ner: 1,
  kex_extract: 25,
  fuse_merge: 10,
  talk_query: 5,
} as const;

export type CreditAction = keyof typeof CREDIT_COSTS;

export function calculateCredits(action: CreditAction, chars: number): number {
  const cost = CREDIT_COSTS[action];
  if (action === 'fuse_merge' || action === 'talk_query') return cost;
  return Math.ceil((chars / 1000) * cost);
}

File: services/agent-rs/src/credits.rs
pub fn calculate(action: &str, chars: u64) -> i64 {
    match action {
        "kex_ner"     => ((chars as f64 / 1000.0) * 1.0).ceil() as i64,
        "kex_extract" => ((chars as f64 / 1000.0) * 25.0).ceil() as i64,
        "fuse_merge"  => 10,
        "talk_query"  => 5,
        _             => 1,
    }
}
